File: src/com/jepl/lang/libraries/arrays/arrays.java
package com.jepl.lang.libraries.arrays;

import com.jepl.lang.Values;
import com.jepl.lang.libraries.Function;
import com.jepl.lang.libraries.Library;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class arrays implements Library {

    @Override
    public HashMap<String, Function> invoke() {
        HashMap<String, Function> out = new HashMap<>();
        out.put("array_new", args -> {
            final String vname = args.get(0);
            Values.addVar(vname, new ArrayList<>());
        });

        out.put("array_put", args -> {
           ArrayList<Object> list = (ArrayList<Object>) Values.vars.get(args.get(0));
           final String value = args.get(1);
           list.add(value);
           Values.addVar(args.get(0), list);
        });

        out.put("array_get", args -> {
            ArrayList<Object> list = (ArrayList<Object>) Values.vars.get(args.get(0));
            final String var = args.get(2);
            final int index = Integer.parseInt(args.get(1));
            Values.addVar(var, list.get(index-1));
        });

        out.put("array_unset", args -> {
            ArrayList<Object> list = (ArrayList<Object>) Values.vars.get(args.get(0));
            final int index = Integer.parseInt(args.get(1));
            list.remove(index-1);
            Values.addVar(args.get(0), list);
        });

        out.put("array_putIndex", args -> {
            ArrayList<Object> list = (ArrayList<Object>) Values.vars.get(args.get(0));
            final int index = Integer.parseInt(args.get(1));
            final String of = args.get(2);
            list.add(index-1, of);
            Values.addVar(args.get(0), list);
        });
        return out;
    }
}

File: src/com/jepl/lang/libraries/arrays/arrays.kt
package com.jepl.lang.libraries.arrays

import com.jepl.lang.Values
import com.jepl.lang.libraries.Function
import com.jepl.lang.libraries.Library

class arrays: Library {
    override fun invoke(): HashMap<String, Function> {
        val out: HashMap<String, Function> = HashMap()

        out["array_new"] = object : Function {
            override fun invoke(args: List<String>) {
                val varName = args[0]
                Values.addVar(varName, arrayListOf<Any>())
            }
        }

        out["array_put"] = object : Function {
            override fun invoke(args: List<String>) {
                val value = args[1]
                (Values.vars[args[0]] as ArrayList<Any>?)?.add(value)
            }
        }

        out["array_get"] = object : Function {
            override fun invoke(args: List<String>) {
                val list = Values.vars[args[0]] as java.util.ArrayList<Any>?
                val `var` = args[2]
                val index = args[1].toInt()
                Values.addVar(`var`, list!![index - 1])
            }
        }

        out["array_unset"] = object : Function {
            override fun invoke(args: List<String>) {
                val list =
                    Values.vars[args[0]] as java.util.ArrayList<Any>?
                val index = args[1].toInt()
                list!!.removeAt(index - 1)
                Values.addVar(args[0], list)
            }
        }

        out["array_putIndex"] = object : Function {
            override fun invoke(args: List<String>) {
                val list =
                    Values.vars[args[0]] as java.util.ArrayList<Any>?
                val index = args[1].toInt()
                val of = args[2]
                list!!.add(index - 1, of)
                Values.addVar(args[0], list)
            }
        }

        return out
    }
}